File: practice-cpp/fundamentals/src/strings.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

int main() {

    // ===== CREACIÓN DE STRINGS =====

    cout << "\n===== CREACIÓN DE STRINGS =====\n";

    string texto1 = "Hola mundo";
    string texto2("Hola desde constructor");
    string texto3(5, 'A'); // 5 veces el carácter 'A'

    cout << "texto1: " << texto1 << "\n";
    cout << "texto2: " << texto2 << "\n";
    cout << "texto3: " << texto3 << "\n";

    // ===== MÉTODO LENGTH() Y SIZE() =====

    cout << "\n===== MÉTODO LENGTH() Y SIZE() =====\n";

    string palabra = "Java";
    int longitud = palabra.length();
    cout << "String: \"" << palabra << "\"\n";
    cout << "Longitud: " << longitud << "\n";

    string frase = "Aprendiendo C++ es divertido";
    cout << "String: \"" << frase << "\"\n";
    cout << "Longitud: " << frase.length() << "\n";

    // ===== MÉTODO FIND(): BUSCAR SUBSTRING =====

    cout << "\n===== MÉTODO FIND(): BUSCAR SUBSTRING =====\n";

    string oracion = "El gato está en la casa";
    cout << "String: \"" << oracion << "\"\n";

    size_t posicion = oracion.find("gato");
    cout << "Posición de \"gato\": " << posicion << "\n";

    size_t posicionNoEncontrada = oracion.find("perro");
    cout << "Posición de \"perro\" (no existe): " << posicionNoEncontrada << "\n";

    // ===== MÉTODO SUBSTR(): EXTRAER SUBSTRING =====

    cout << "\n===== MÉTODO SUBSTR(): EXTRAER SUBSTRING =====\n";

    string frase2 = "Programación en C++";
    cout << "String original: \"" << frase2 << "\"\n";

    string parte1 = frase2.substr(0, 12); // desde posición 0, 12 caracteres
    cout << "substr(0, 12): \"" << parte1 << "\"\n";

    string parte2 = frase2.substr(17, 3); // desde posición 17, 3 caracteres
    cout << "substr(17, 3): \"" << parte2 << "\"\n";

    // ===== MÉTODO REPLACE(): REEMPLAZAR =====

    cout << "\n===== MÉTODO REPLACE(): REEMPLAZAR =====\n";

    string original = "Hola Mundo";
    cout << "String original: \"" << original << "\"\n";

    original.replace(5, 5, "C++");
    cout << "Después de replace(5, 5, \"C++\"): \"" << original << "\"\n";

    // ===== MÉTODOS TOUPPERCASE() Y TOLOWERCASE() =====

    cout << "\n===== TRANSFORM A MAYÚSCULAS/MINÚSCULAS =====\n";

    string mixto = "PrOgRaMaCióN";
    cout << "String original: \"" << mixto << "\"\n";

    string mayusculas = mixto;
    transform(mayusculas.begin(), mayusculas.end(), mayusculas.begin(), ::toupper);
    cout << "Mayúsculas: \"" << mayusculas << "\"\n";

    string minusculas = mixto;
    transform(minusculas.begin(), minusculas.end(), minusculas.begin(), ::tolower);
    cout << "Minúsculas: \"" << minusculas << "\"\n";

    // ===== MÉTODO COMPARE(): COMPARAR STRINGS =====

    cout << "\n===== MÉTODO COMPARE(): COMPARAR STRINGS =====\n";

    string a = "Apple";
    string b = "Banana";
    string c = "Apple";

    cout << "\"" << a << "\".compare(\"" << b << "\"): " << a.compare(b) << " (negativo si a < b)\n";
    cout << "\"" << b << "\".compare(\"" << a << "\"): " << b.compare(a) << " (positivo si b > a)\n";
    cout << "\"" << a << "\".compare(\"" << c << "\"): " << a.compare(c) << " (cero si son iguales)\n";

    // ===== OPERADOR == PARA COMPARACIÓN =====

    cout << "\n===== OPERADOR == PARA COMPARACIÓN =====\n";

    string password = "Seguro123";
    string passwordIngresado = "Seguro123";
    string passwordIncorrecto = "seguro123";

    cout << "password == passwordIngresado: " << (password == passwordIngresado ? "true" : "false") << "\n";
    cout << "password == passwordIncorrecto: " << (password == passwordIncorrecto ? "true" : "false") << "\n";

    // ===== MÉTODO APPEND(): CONCATENAR =====

    cout << "\n===== MÉTODO APPEND(): CONCATENAR =====\n";

    string frase3 = "Hola";
    cout << "Antes: \"" << frase3 << "\"\n";

    frase3.append(" ");
    frase3.append("Mundo");
    cout << "Después: \"" << frase3 << "\"\n";

    // ===== OPERADOR + PARA CONCATENACIÓN =====

    cout << "\n===== OPERADOR + PARA CONCATENACIÓN =====\n";

    string saludo = "Buenos " + string("días");
    cout << "Concatenado: \"" << saludo << "\"\n";

    // ===== MÉTODO TRIM() (ELIMINAR ESPACIOS AL INICIO Y FINAL) =====

    cout << "\n===== TRIM: ELIMINAR ESPACIOS AL INICIO Y FINAL =====\n";

    string conEspacios = "  Hola Mundo  ";
    cout << "String original: \"" << conEspacios << "\"\n";
    cout << "Longitud: " << conEspacios.length() << "\n";

    // En C++ no hay trim nativo, pero podemos usar substr
    size_t inicio = conEspacios.find_first_not_of(" ");
    size_t fin = conEspacios.find_last_not_of(" ");
    string trimmeado = conEspacios.substr(inicio, fin - inicio + 1);
    cout << "Después de trim: \"" << trimmeado << "\"\n";
    cout << "Longitud: " << trimmeado.length() << "\n";

    // ===== MÉTODO SPLIT() (DIVIDIR STRING) =====

    cout << "\n===== SPLIT: DIVIDIR STRING (SIMULADO) =====\n";

    string csv = "Juan,Carlos,Diana,Bruno";
    cout << "String original: \"" << csv << "\"\n";

    cout << "Dividido por coma:\n";
    stringstream ss(csv);
    string item;
    while (getline(ss, item, ',')) {
        cout << "  - " << item << "\n";
    }

    // ===== CONVERSIÓN A NÚMERO: STOI(), STOD() =====

    cout << "\n===== CONVERSIÓN A NÚMERO =====\n";

    string numeroStr = "42";
    int numeroInt = stoi(numeroStr);
    cout << "String \"" << numeroStr << "\" a int: " << numeroInt << "\n";

    string decimalStr = "3.14159";
    double numeroDouble = stod(decimalStr);
    cout << "String \"" << decimalStr << "\" a double: " << numeroDouble << "\n";

    // ===== CONVERSIÓN A STRING: TO_STRING() =====

    cout << "\n===== CONVERSIÓN A STRING =====\n";

    int numero = 42;
    double decimal = 3.14;
    string stringNumero = to_string(numero);
    string stringDecimal = to_string(decimal);

    cout << "int " << numero << " a string: \"" << stringNumero << "\"\n";
    cout << "double " << decimal << " a string: \"" << stringDecimal << "\"\n";

    // ===== GETLINE(): LEER LÍNEA COMPLETA (COMENTADO PARA DEMO) =====

    cout << "\n===== GETLINE(): LEER LÍNEA COMPLETA =====\n";
    cout << "(Saltado en esta demo para no bloquear la ejecución)\n";
    // cout << "Ingrese una línea: ";
    // string linea;
    // getline(cin, linea);
    // cout << "Usted ingresó: \"" << linea << "\"\n";

    return 0;
}

File: practice-java/fundamentals/src/Strings.java
public class Strings {
    public static void main(String[] args) {

        // ===== CREACIÓN DE STRINGS =====

        System.out.println("===== CREACIÓN DE STRINGS =====");

        // forma 1: literal de string
        String mensaje = "Hola mundo";

        // forma 2: usando constructor (menos común)
        String mensaje2 = new String("Hola desde constructor");

        System.out.println(mensaje);
        System.out.println(mensaje2);

        // ===== MÉTODO LENGTH(): LONGITUD DEL STRING =====

        System.out.println("\n===== MÉTODO LENGTH() =====");

        String texto = "Java";
        int longitud = texto.length();
        System.out.println("String: \"" + texto + "\"");
        System.out.println("Longitud: " + longitud);

        String frase = "Aprendiendo Java es divertido";
        System.out.println("String: \"" + frase + "\"");
        System.out.println("Longitud: " + frase.length());

        // ===== MÉTODO CHARAT(): OBTENER CARÁCTER POR ÍNDICE =====

        System.out.println("\n===== MÉTODO CHARAT() =====");

        String palabra = "Python";
        System.out.println("String: \"" + palabra + "\"");
        System.out.println("Carácter en índice 0: " + palabra.charAt(0));
        System.out.println("Carácter en índice 2: " + palabra.charAt(2));
        System.out.println("Carácter en índice 5 (último): " + palabra.charAt(5));

        // ===== MÉTODO SUBSTRING(): EXTRAER PARTE DEL STRING =====

        System.out.println("\n===== MÉTODO SUBSTRING() =====");

        String frase2 = "Programación en Java";
        System.out.println("String original: \"" + frase2 + "\"");

        // substring(inicio) - desde la posición inicio hasta el final
        String parte1 = frase2.substring(0, 11);
        System.out.println("substring(0, 11): \"" + parte1 + "\"");

        // substring(inicio, fin) - desde inicio hasta fin (fin no incluido)
        String parte2 = frase2.substring(15, 19);
        System.out.println("substring(15, 19): \"" + parte2 + "\"");

        // ===== MÉTODO INDEXOF(): BUSCAR POSICIÓN DE CARÁCTER =====

        System.out.println("\n===== MÉTODO INDEXOF() =====");

        String oracion = "El gato está en la casa";
        System.out.println("String: \"" + oracion + "\"");

        int posicion = oracion.indexOf("gato");
        System.out.println("Posición de \"gato\": " + posicion);

        int posicionEspacio = oracion.indexOf(" ");
        System.out.println("Posición del primer espacio: " + posicionEspacio);

        int noEncontrado = oracion.indexOf("perro");
        System.out.println("Posición de \"perro\" (no existe): " + noEncontrado); // retorna -1

        // ===== MÉTODO CONTAINS(): VERIFICAR SI CONTIENE SUBSTRING =====

        System.out.println("\n===== MÉTODO CONTAINS() =====");

        String texto2 = "Me encanta programar en Java";
        System.out.println("String: \"" + texto2 + "\"");
        System.out.println("¿Contiene 'Java'? " + texto2.contains("Java"));
        System.out.println("¿Contiene 'Python'? " + texto2.contains("Python"));
        System.out.println("¿Contiene 'programar'? " + texto2.contains("programar"));

        // ===== MÉTODO REPLACE(): REEMPLAZAR CARACTERES =====

        System.out.println("\n===== MÉTODO REPLACE() =====");

        String original = "Hola Mundo";
        System.out.println("String original: \"" + original + "\"");

        String reemplazado = original.replace("Mundo", "Java");
        System.out.println("Después de reemplazar 'Mundo' por 'Java': \"" + reemplazado + "\"");

        String conEspacios = "Hola Mundo Hola";
        String sinHolas = conEspacios.replace("Hola", "Hey");
        System.out.println("Reemplazar todas las ocurrencias: \"" + sinHolas + "\"");

        // ===== MÉTODOS TOUPPERCASE() Y TOLOWERCASE() =====

        System.out.println("\n===== MÉTODOS TOUPPERCASE() Y TOLOWERCASE() =====");

        String mixto = "PrOgRaMaCiÓn";
        System.out.println("String original: \"" + mixto + "\"");
        System.out.println("toUpperCase(): \"" + mixto.toUpperCase() + "\"");
        System.out.println("toLowerCase(): \"" + mixto.toLowerCase() + "\"");

        // ===== MÉTODO TRIM(): ELIMINAR ESPACIOS AL INICIO Y FINAL =====

        System.out.println("\n===== MÉTODO TRIM() =====");

        String conEspacioEnBlanco = "  Hola Mundo  ";
        System.out.println("String con espacios: \"" + conEspacioEnBlanco + "\"");
        System.out.println("Longitud original: " + conEspacioEnBlanco.length());

        String sinEspacios = conEspacioEnBlanco.trim();
        System.out.println("Después de trim(): \"" + sinEspacios + "\"");
        System.out.println("Longitud después de trim: " + sinEspacios.length());

        // ===== MÉTODO EQUALS(): COMPARAR STRINGS =====

        System.out.println("\n===== MÉTODO EQUALS() =====");

        String password = "Seguro123";
        String passwordIngresado = "Seguro123";
        String passwordIncorrecto = "seguro123";

        System.out.println("password.equals(passwordIngresado): " + password.equals(passwordIngresado));
        System.out.println("password.equals(passwordIncorrecto): " + password.equals(passwordIncorrecto));

        // equalsIgnoreCase(): ignora mayúsculas/minúsculas
        System.out.println("password.equalsIgnoreCase(passwordIncorrecto): " + password.equalsIgnoreCase(passwordIncorrecto));

        // ===== MÉTODO COMPARETO(): COMPARACIÓN LEXICOGRÁFICA =====

        System.out.println("\n===== MÉTODO COMPARETO() =====");

        String a = "Apple";
        String b = "Banana";
        String c = "Apple";

        System.out.println("\"" + a + "\".compareTo(\"" + b + "\"): " + a.compareTo(b)); // negativo (A < B)
        System.out.println("\"" + b + "\".compareTo(\"" + a + "\"): " + b.compareTo(a)); // positivo (B > A)
        System.out.println("\"" + a + "\".compareTo(\"" + c + "\"): " + a.compareTo(c)); // cero (igual)

        // ===== MÉTODO SPLIT(): DIVIDIR STRING EN ARRAY =====

        System.out.println("\n===== MÉTODO SPLIT() =====");

        String csv = "Juan,Carlos,Diana,Bruno";
        System.out.println("String original: \"" + csv + "\"");

        String[] nombres = csv.split(",");
        System.out.println("Dividido por coma:");
        for (String nombre : nombres) {
            System.out.println("  - " + nombre);
        }

        String frase3 = "El lenguaje Java es poderoso";
        String[] palabras = frase3.split(" ");
        System.out.println("\n\"" + frase3 + "\"");
        System.out.println("Dividido por espacios:");
        for (String palabra2 : palabras) {
            System.out.println("  - " + palabra2);
        }

        // ===== INMUTABILIDAD DE STRINGS =====

        System.out.println("\n===== INMUTABILIDAD DE STRINGS =====");

        // Los strings en Java son inmutables
        String original2 = "Hola";
        System.out.println("String original: \"" + original2 + "\"");

        String modificado = original2.toUpperCase();
        System.out.println("Después de toUpperCase(): \"" + modificado + "\"");
        System.out.println("Original sigue siendo: \"" + original2 + "\"");
        System.out.println("¿original2 == \"Hola\"? " + original2.equals("Hola")); // true

        // ===== STRINGBUILDER: CONCATENACIÓN EFICIENTE =====

        System.out.println("\n===== STRINGBUILDER: CONCATENACIÓN EFICIENTE =====");

        // Concatenación con + (menos eficiente para bucles)
        String resultado1 = "";
        for (int i = 1; i <= 5; i++) {
            resultado1 += i;
        }
        System.out.println("Concatenación con +: " + resultado1);

        // StringBuilder (más eficiente para múltiples concatenaciones)
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= 5; i++) {
            sb.append(i);
        }
        String resultado2 = sb.toString();
        System.out.println("Concatenación con StringBuilder: " + resultado2);

        // StringBuilder ejemplo práctico
        StringBuilder fraseBuild = new StringBuilder();
        fraseBuild.append("Me llamo ");
        fraseBuild.append("Fernando");
        fraseBuild.append(" y me encanta ");
        fraseBuild.append("programar en Java");
        System.out.println("Frase construida: " + fraseBuild.toString());

        // ===== CONVERSIÓN A STRING =====

        System.out.println("\n===== CONVERSIÓN A STRING =====");

        int numero = 42;
        String stringNumero = String.valueOf(numero);
        System.out.println("int " + numero + " convertido a String: \"" + stringNumero + "\"");

        double decimal = 3.14;
        String stringDecimal = String.valueOf(decimal);
        System.out.println("double " + decimal + " convertido a String: \"" + stringDecimal + "\"");

    }
}
